agg-ast/schema_derivation/src/schema_derivation.rs

```diff
@@ -35,21 +37,210 @@ pub(crate) struct ResultSetState<'a> {
+                        Some(schema) => Some(schema.union(&doc_schema)),
+                    })
+                },
+            );
+            Ok(schema?.unwrap_or(Schema::Document(Document::default())))
+        }
+
+        fn facet_derive_schema(
+            facet: &LinkedHashMap<String, Vec<Stage>>,
+            state: &mut ResultSetState,
+        ) -> Result<Schema> {
+            // facet contains key - value pairs where the key is the name of a field in the output document,
+            // and the value is an aggregation pipeline. We can use the same generic helper for aggregating over a whole pipeline
+            // to get the sch ema for each field, cloning the incoming state.
+            let facet_schema = facet
+                .into_iter()
+                .map(|(field, pipeline)| {
+                    let mut field_state = state.clone();
+                    let field_schema =
+                        derive_schema_for_pipeline(pipeline.clone(), &mut field_state)?;
+                    Ok((field.clone(), field_schema))
+                })
+                .collect::<Result<BTreeMap<String, Schema>>>()?;
+            Ok(Schema::Document(Document {
+                required: facet_schema.keys().cloned().collect(),
+                keys: facet_schema,
+                ..Default::default()
+            }))
+        }
+
+        fn sort_by_count_derive_schema(
+            sort_expr: &Expression,
+            state: &mut ResultSetState,
+        ) -> Result<Schema> {
+            Ok(Schema::Document(Document {
+                keys: map! {
+                    "_id".to_string() => sort_expr.derive_schema(state)?,
+                    "count".to_string() => Schema::AnyOf(set!(Schema::Atomic(Atomic::Integer), Schema::Atomic(Atomic::Long)))
+                },
+                required: set!("_id".to_string(), "count".to_string()),
+                ..Default::default()
+            }))
+        }
+
+        fn unwind_derive_schema(unwind: &Unwind, state: &mut ResultSetState) -> Result<Schema> {
+            let path = match unwind {
+                Unwind::FieldPath(Expression::Ref(Ref::FieldRef(r))) => {
+                    Some(r.split(".").map(|s| s.to_string()).collect::<Vec<String>>())
+                }
+                Unwind::Document(d) => {
+                    if let Expression::Ref(Ref::FieldRef(r)) = d.path.as_ref() {
+                        Some(r.split(".").map(|s| s.to_string()).collect::<Vec<String>>())
+                    } else {
+                        None
+                    }
+                }
+                _ => None,
+            };
+            if let Some(path) = path {
+                if let Some(s) = get_schema_for_path_mut(&mut state.result_set_schema, path) {
+                    // the schema of the field being unwound goes from type Array[X] to type X
+                    match s {
+                        Schema::Array(a) => {
+                            *s = *a.clone();
```

Comment:
We can do an `std::mem::take` for sure here.

Reply:
Oh, we may not have default defined for Schema. I forget. If not, just derive Default for Schema with Unsat at the Default.